File: clients/file-reader/file_reader/smartedu/metadata.py
# Mirrors chrome-extension/doc-extract/smartedu/metadata.js — keep ti_items walk rules in sync.

"""Fetch SmartEdu detail JSON and extract downloadable assets."""

from __future__ import annotations

import json
from typing import Any, Callable, Optional
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from file_reader.smartedu.models import SmartEduAsset, SmartEduLesson
from file_reader.smartedu.url_parser import ParsedSmartEduUrl, SmartEduUrlKind

M3U8_FLAGS = ("href-m3u8", "href")
PDF_FLAG = "pdf"
TARGET_TYPES = (
    "micro_lesson_video",
    "coursewares",
    "lesson_plandesign",
    "learning_task",
)
M3U8_RESOLUTION_ORDER = ("1920x1080", "1280x720", "852x480", "640x360")


def _localized_title(value: Any) -> str:
    if isinstance(value, dict):
        for key in ("zh-CN", "zh_CN", "zh"):
            text = value.get(key)
            if isinstance(text, str) and text.strip():
                return text.strip()
    if isinstance(value, str) and value.strip():
        return value.strip()
    return ""


def _custom_props(resource: dict[str, Any]) -> dict[str, Any]:
    props = resource.get("custom_properties")
    return props if isinstance(props, dict) else {}


def _pick_m3u8_item(items: list[dict[str, Any]]) -> Optional[dict[str, Any]]:
    candidates: list[tuple[int, dict[str, Any]]] = []
    for item in items:
        if not isinstance(item, dict):
            continue
        flag = str(item.get("ti_file_flag") or "")
        fmt = str(item.get("ti_format") or "")
        if fmt != "m3u8" or flag not in M3U8_FLAGS:
            continue
        storages = item.get("ti_storages")
        if not isinstance(storages, list) or not storages:
            continue
        url = str(storages[0] or "")
        if not url:
            continue
        rank = 0
        if flag == "href-m3u8":
            rank += 100
        for index, token in enumerate(M3U8_RESOLUTION_ORDER):
            if token in url:
                rank += (len(M3U8_RESOLUTION_ORDER) - index) * 10
                break
        candidates.append((rank, item))
    if not candidates:
        return None
    candidates.sort(key=lambda pair: pair[0], reverse=True)
    return candidates[0][1]


def _pick_pdf_item(items: list[dict[str, Any]]) -> Optional[dict[str, Any]]:
    for item in items:
        if not isinstance(item, dict):
            continue
        if str(item.get("ti_file_flag") or "") != PDF_FLAG:
            continue
        storages = item.get("ti_storages")
        if isinstance(storages, list) and storages:
            return item
    return None


def _storage_url(item: dict[str, Any]) -> str:
    storages = item.get("ti_storages")
    if not isinstance(storages, list):
        return ""
    for entry in storages:
        text = str(entry or "").strip()
        if text:
            return text
    return ""


def _asset_from_resource(resource: dict[str, Any]) -> Optional[SmartEduAsset]:
    resource_type = str(resource.get("resource_type_code") or "")
    if resource_type not in TARGET_TYPES:
        return None

    props = _custom_props(resource)
    alias = str(props.get("alias_name") or resource.get("title") or resource_type)
    title = _localized_title(resource.get("global_title")) or str(resource.get("title") or alias)

    items = resource.get("ti_items")
    if not isinstance(items, list):
        return None

    if resource_type == "micro_lesson_video":
        picked = _pick_m3u8_item(items)
        if picked is None:
            return None
        download_url = _storage_url(picked)
        file_format = "mp4"
    else:
        picked = _pick_pdf_item(items)
        if picked is None:
            return None
        download_url = _storage_url(picked)
        file_format = "pdf"

    if not download_url:
        return None

    asset_id = str(resource.get("id") or resource.get("version_id") or title)
    return SmartEduAsset(
        asset_id=asset_id,
        title=title,
        alias=alias,
        resource_type=resource_type,
        format=file_format,
        download_url=download_url,
    )


def extract_assets_from_detail(detail: dict[str, Any]) -> list[SmartEduAsset]:
    """Walk relations.* and return the four standard lesson assets."""
    relations = detail.get("relations")
    if not isinstance(relations, dict):
        return []

    resources = relations.get("national_course_resource")
    if not isinstance(resources, list):
        return []

    assets: list[SmartEduAsset] = []
    seen_types: set[str] = set()
    for resource in resources:
        if not isinstance(resource, dict):
            continue
        asset = _asset_from_resource(resource)
        if asset is None:
            continue
        if asset.resource_type in seen_types:
            continue
        seen_types.add(asset.resource_type)
        assets.append(asset)

    order = {name: index for index, name in enumerate(TARGET_TYPES)}
    assets.sort(key=lambda item: order.get(item.resource_type, 99))
    return assets


def lesson_title_from_detail(detail: dict[str, Any]) -> str:
    """Resolve display title from detail JSON."""
    title = _localized_title(detail.get("global_title"))
    if title:
        return title
    return str(detail.get("title") or detail.get("id") or "SmartEdu lesson")


def build_lesson(parsed: ParsedSmartEduUrl, detail: dict[str, Any]) -> SmartEduLesson:
    """Combine parse result and detail JSON into a lesson model."""
    lesson_id = str(detail.get("id") or parsed.resource_id)
    return SmartEduLesson(
        lesson_id=lesson_id,
        title=lesson_title_from_detail(detail),
        detail_url=parsed.detail_url,
        assets=extract_assets_from_detail(detail),
    )


def fetch_detail_json(
    detail_url: str,
    *,
    timeout: int = 30,
    opener: Optional[Callable[..., Any]] = None,
) -> dict[str, Any]:
    """GET detail JSON from SmartEdu CDN."""
    request = Request(
        detail_url,
        headers={
            "Accept": "application/json",
            "User-Agent": "MindGraph-FileReader/1.0",
        },
        method="GET",
    )
    open_fn = opener or urlopen
    try:
        with open_fn(request, timeout=timeout) as response:
            raw = response.read().decode("utf-8")
    except HTTPError as exc:
        body = exc.read().decode("utf-8", errors="replace")
        raise ValueError(f"HTTP {exc.code}: {body[:200]}") from exc
    except URLError as exc:
        raise ValueError(f"Network error: {exc.reason}") from exc

    try:
        data = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ValueError("Detail response is not valid JSON") from exc
    if not isinstance(data, dict):
        raise ValueError("Detail response must be a JSON object")
    return data


def load_lesson_from_detail_file(path: str) -> SmartEduLesson:
    """Load a lesson from a local fixture or cached detail JSON."""
    with open(path, encoding="utf-8") as handle:
        detail = json.load(handle)
    if not isinstance(detail, dict):
        raise ValueError("Fixture must be a JSON object")
    lesson_id = str(detail.get("id") or "fixture")
    parsed = ParsedSmartEduUrl(
        kind=SmartEduUrlKind.CLASS_ACTIVITY,
        resource_id=lesson_id,
        detail_url=f"fixture://{path}",
    )
    return build_lesson(parsed, detail)


def fetch_lesson(parsed: ParsedSmartEduUrl, *, timeout: int = 30) -> SmartEduLesson:
    """Fetch detail JSON and extract assets."""
    detail = fetch_detail_json(parsed.detail_url, timeout=timeout)
    return build_lesson(parsed, detail)

File: chrome-extension/doc-extract/smartedu/metadata.js
/**
 * SmartEdu metadata fetch + ti_items walk — mirrors file_reader/smartedu/metadata.py.
 * Keep ti_items walk rules in sync with the file-reader SmartEdu tab.
 * @fileoverview Flatten lesson assets from detail JSON.
 */
(function (global) {
  "use strict";

  const MindGraphDocExtract = global.MindGraphDocExtract || {};
  const createSmartEduAsset = MindGraphDocExtract.createSmartEduAsset;

  const ALIAS_LABELS = {
    coursewares: "课件",
    lesson_plandesign: "教学设计",
    learning_task: "学习任务单",
  };

  /**
   * @param {string} alias
   * @param {{ localKind?: string }} picked
   * @returns {boolean}
   */
  function isDocumentAsset(alias, picked) {
    if (alias === "micro_lesson_video") {
      return false;
    }
    if (picked && picked.localKind === "m3u8") {
      return false;
    }
    return true;
  }

  /**
   * Pick best download URL from one ti_item (storages may be URL strings or objects).
   * @param {object} item
   * @returns {{ url: string, format: string, localKind: string, rank: number }}
   */
  function pickDownloadFromTiItem(item) {
    const storages = Array.isArray(item.ti_storages) ? item.ti_storages : [];
    const files = Array.isArray(item.ti_files) ? item.ti_files : [];
    const candidates = [];
    const itemFlag = String(item.ti_file_flag || item.ti_format || "").toLowerCase();

    storages.forEach((s) => {
      if (typeof s === "string" && s.trim()) {
        const href = s.trim();
        const flag = itemFlag || (href.includes(".m3u8") ? "m3u8" : "");
        candidates.push(buildCandidate(href, flag));
        return;
      }
      if (!s || typeof s !== "object") {
        return;
      }
      const href = s["href-m3u8"] || s.href || s.url || "";
      if (!href) {
        return;
      }
      const flag = String(s.ti_file_flag || s.format || itemFlag || "").toLowerCase();
      candidates.push(buildCandidate(String(href), flag));
    });

    files.forEach((f) => {
      if (!f || typeof f !== "object") {
        return;
      }
      const href = f.href || f.url || "";
      if (!href) {
        return;
      }
      const flag = String(f.ti_file_flag || f.format || itemFlag || "").toLowerCase();
      candidates.push(buildCandidate(String(href), flag));
    });

    if (!candidates.length) {
      return { url: "", format: "", localKind: "file", rank: 0 };
    }
    candidates.sort((a, b) => b.rank - a.rank);
    return candidates[0];
  }

  /**
   * @param {string} href
   * @param {string} flag
   * @returns {{ url: string, format: string, localKind: string, rank: number }}
   */
  function buildCandidate(href, flag) {
    const lowerFlag = (flag || "").toLowerCase();
    let localKind = "file";
    let rank = 1;
    if (lowerFlag === "pdf" || href.toLowerCase().includes(".pdf")) {
      localKind = "pdf";
      rank = 100;
    } else if (lowerFlag.includes("m3u8") || href.toLowerCase().includes(".m3u8")) {
      localKind = "m3u8";
      rank = lowerFlag === "href-m3u8" ? 90 : lowerFlag.includes("1080") ? 85 : 50;
    }
    return {
      url: href,
      format: lowerFlag || localKind,
      localKind,
      rank,
    };
  }

  /**
   * Pick one asset URL from a national_course_resource block (all ti_items).
   * @param {object} block
   * @returns {{ url: string, format: string, localKind: string }}
   */
  function pickBestFromResourceBlock(block) {
    const tiItems = block && Array.isArray(block.ti_items) ? block.ti_items : [];
    let best = { url: "", format: "", localKind: "file", rank: 0 };
    tiItems.forEach((item) => {
      if (!item || typeof item !== "object") {
        return;
      }
      const picked = pickDownloadFromTiItem(item);
      if (picked.url && picked.rank >= best.rank) {
        best = picked;
      }
    });
    return best;
  }

  /**
   * @param {object} block
   * @returns {string}
   */
  function blockTitle(block) {
    if (block.title) {
      return String(block.title);
    }
    const gt = block.global_title;
    if (gt && typeof gt === "object") {
      return String(gt["zh-CN"] || gt.zh_cn || gt.en || Object.values(gt)[0] || "");
    }
    return "";
  }

  /**
   * @param {object} block
   * @returns {string}
   */
  function blockAlias(block) {
    if (block.resource_type_code) {
      return String(block.resource_type_code);
    }
    if (block.custom_properties && block.custom_properties.alias_name) {
      return String(block.custom_properties.alias_name);
    }
    return "asset";
  }

  /**
   * @param {object} detailJson
   * @returns {Array<object>}
   */
  function extractAssetsFromDetailJson(detailJson) {
    const assets = [];
    const relations = detailJson && detailJson.relations ? detailJson.relations : {};

    Object.keys(relations).forEach((relKey) => {
      const bucket = relations[relKey];
      const items = Array.isArray(bucket) ? bucket : bucket && typeof bucket === "object" ? [bucket] : [];
      items.forEach((block) => {
        if (!block || typeof block !== "object") {
          return;
        }
        const alias = blockAlias(block);
        const picked = pickBestFromResourceBlock(block);
        if (!picked.url || !isDocumentAsset(alias, picked)) {
          return;
        }
        const label = blockTitle(block) || ALIAS_LABELS[alias] || alias;
        assets.push(
          createSmartEduAsset({
            id: String(block.id || block.version_id || `${alias}-${assets.length}`),
            title: label,
            alias,
            format: picked.format,
            downloadUrl: picked.url,
            localKind: picked.localKind,
            selected: true,
          }),
        );
      });
    });

    if (!assets.length && detailJson && detailJson.ti_items) {
      const tiItems = Array.isArray(detailJson.ti_items) ? detailJson.ti_items : [];
      tiItems.forEach((item, idx) => {
        const picked = pickDownloadFromTiItem(item);
        if (!picked.url || !isDocumentAsset(alias, picked)) {
          return;
        }
        const alias = String(item.alias || item.ti_type || `item_${idx}`);
        assets.push(
          createSmartEduAsset({
            id: String(item.id || `root-${idx}`),
            title: String(item.title || ALIAS_LABELS[alias] || alias),
            alias,
            format: picked.format,
            downloadUrl: picked.url,
            localKind: picked.localKind,
            selected: true,
          }),
        );
      });
    }

    return assets;
  }

  /**
   * @param {string} detailUrl
   * @param {Record<string, string>} headers
   * @returns {Promise<{ title: string, assets: Array<object> }>}
   */
  async function fetchSmartEduMetadata(detailUrl, headers) {
    const res = await fetch(detailUrl, {
      method: "GET",
      headers: {
        Accept: "application/json",
        ...headers,
      },
    });
    if (!res.ok) {
      if (res.status === 401 || res.status === 403) {
        await MindGraphDocExtract.clearSmartEduToken();
      }
      throw new Error(`SmartEdu metadata HTTP ${res.status}`);
    }
    const detailJson = await res.json();
    const title =
      (detailJson && (detailJson.title || detailJson.global_title?.["zh-CN"] ||
        detailJson.global_title?.zh_cn)) || "SmartEdu lesson";
    return {
      title: String(title),
      assets: extractAssetsFromDetailJson(detailJson),
    };
  }

  MindGraphDocExtract.extractAssetsFromDetailJson = extractAssetsFromDetailJson;
  MindGraphDocExtract.fetchSmartEduMetadata = fetchSmartEduMetadata;
  MindGraphDocExtract.SMARTEDU_ALIAS_LABELS = ALIAS_LABELS;
  global.MindGraphDocExtract = MindGraphDocExtract;
})(typeof self !== "undefined" ? self : globalThis);
